File: modal-go/proxy.go
package modal

import (
	"context"
	"fmt"

	pb "github.com/modal-labs/libmodal/modal-go/proto/modal_proto"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// ProxyService provides Proxy related operations.
type ProxyService struct{ client *Client }

// Proxy represents a Modal Proxy.
type Proxy struct {
	ProxyId string
}

// ProxyFromNameOptions are options for looking up a Modal Proxy.
type ProxyFromNameOptions struct {
	Environment string
}

// FromName references a modal.Proxy by its name.
func (s *ProxyService) FromName(ctx context.Context, name string, options *ProxyFromNameOptions) (*Proxy, error) {
	if options == nil {
		options = &ProxyFromNameOptions{}
	}

	resp, err := s.client.cpClient.ProxyGet(ctx, pb.ProxyGetRequest_builder{
		Name:            name,
		EnvironmentName: environmentName(options.Environment, s.client.profile),
	}.Build())

	if status, ok := status.FromError(err); ok && status.Code() == codes.NotFound {
		return nil, NotFoundError{fmt.Sprintf("Proxy '%s' not found", name)}
	}
	if err != nil {
		return nil, err
	}

	if resp.GetProxy() == nil || resp.GetProxy().GetProxyId() == "" {
		return nil, NotFoundError{fmt.Sprintf("Proxy '%s' not found", name)}
	}

	return &Proxy{ProxyId: resp.GetProxy().GetProxyId()}, nil
}

File: modal-js/src/proxy.ts
import { getDefaultClient, type ModalClient } from "./client";
import { ClientError, Status } from "nice-grpc";
import { NotFoundError } from "./errors";

/**
 * Service for managing Proxies.
 */
export class ProxyService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Reference a Proxy by its name.
   */
  async fromName(name: string, params?: ProxyFromNameParams): Promise<Proxy> {
    try {
      const resp = await this.#client.cpClient.proxyGet({
        name,
        environmentName: this.#client.environmentName(params?.environment),
      });
      if (!resp.proxy?.proxyId) {
        throw new NotFoundError(`Proxy '${name}' not found`);
      }
      return new Proxy(resp.proxy.proxyId);
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(`Proxy '${name}' not found`);
      throw err;
    }
  }
}

/** Optional parameters for `client.proxies.fromName()`. */
export type ProxyFromNameParams = {
  environment?: string;
};

/** Proxy objects give your Modal containers a static outbound IP address. */
export class Proxy {
  readonly proxyId: string;

  /** @ignore */
  constructor(proxyId: string) {
    this.proxyId = proxyId;
  }

  /**
   * @deprecated Use `client.proxies.fromName()` instead.
   */
  static async fromName(
    name: string,
    params?: ProxyFromNameParams,
  ): Promise<Proxy> {
    return getDefaultClient().proxies.fromName(name, params);
  }
}
